### ISSUE TITLE
[BUG] Addresses (shipping / billing) not validated during addressing

### ISSUE THREAD

**Issue by AndreasA** [Sylius/ShopApiPlugin#588]
Hi,

currently the shipping / billing addresses of a cart are not validated during addressing.
You have code in it that is supposed to do it but it does not work becuase:
- In `Sylius\ShopApiPlugin\Controller\Checkout\AddressAction` you validate the `Sylius\ShopApiPlugin\Request\Checkout\AddressOrderRequest`
- In `Resources/config/validation/checkout/AddressOrderRequest.xml` you use `Valid` constraint for `shippingAddress` and `billingAddress` which in theory is fine but does not work due the last point
- In `Sylius\ShopApiPlugin\Request\Checkout\AddressOrderRequest` the shippingAddress and `billingAddress` is still an array, so `Valid` will not trigger the normal `Address` entity validation. 

You could already pre-parse the address in the constructor of `Sylius\ShopApiPlugin\Request\Checkout\AddressOrderRequest` , e.g. using `Address::createFromArray. However, you would need to allow it to provide invalid data. 

Furthermore, it being a static method makes it difficult to add custom address properties (at this point).

**Comment by mamazu** [comment]
Yes, that's true. One possible solution would be to create the address object in the constructor but this would throw a 500 if properties don't exist. So the from array would need to leave out all non existing properties, which makes it possible to create an address with empty fields.

On the other hand you are completely correct that this static method makes it hard to add more properties to the address. We might also want to add a mapping method to the Address mapper.

**Comment by AndreasA** [comment]
Not perfect but a first temporary solution would be to allow non-set values in:
https://github.com/Sylius/ShopApiPlugin/blob/master/src/Model/Address.php#L66

and use `?? null` when setting the properties. Probably not perfect but at least the validation would work.

Regarding the static method. Probably a service would be better as that is adjustable. The same for "createFromRequest" or an event listener or something similar.

**Comment by mamazu** [comment]
Well the idea behind the assertion is that you can not create an object without those fields being set so you can rely on the fact that they are not empty. One other thing we might want to do is define the validation separately so that we can apply it to the raw data.

With the Address, it is a bit of a special case in the API this is as far as I know the only entity from the API so there is not enough consistency in how to handle that.

**Comment by AndreasA** [comment]
I thought the model in between you use for the ShopPluginApi instead of the Entity is to pre-parse the data and use it for validation? If you expect it to already contain valid data, you could just use the entity directly. But appareantly not as it always happens too late to be used that way.

I guess this means that in most places the address validation will not work and not just checkout? Ii didn't test it elsewhere yet.

**Comment by mamazu** [comment]
It should work with the address book (the validation of the entity is handled elsewhere) and there no other place where the address is used.

**Comment by hatemBBN** [comment]
@mamazu @AndreasA I am sorry, I read the conversation but could not understand why there is another Address model in the shop API ?
and is there any recommended workaround for this issue without side effects ?

**Comment by mamazu** [comment]
Currently there is no workaround. Except for building something completely different and not using the command provider.

**Comment by hatemBBN** [comment]
@mamazu I followed @AndreasA suggestion to create the address object in the constructor
```
$this->shippingAddress = AddressAdapter::createFromArray($request->request->get('shippingAddress'));
```
and used Adapter to set values to empty string when missing to prevent 500 if properties don't exist
```
class AddressAdapter
{
    public static function createFromArray(array $address): Address
    {
        $address['firstName'] = $address['firstName'] ?? '';
        $address['lastName'] = $address['lastName'] ?? '';
        $address['city'] = $address['city'] ?? '';
        $address['street'] = $address['street'] ?? '';
        $address['countryCode'] = $address['countryCode'] ?? '';
        $address['postcode'] = $address['postcode'] ?? '';

        return Address::createFromArray($address);
    }
}
```
and validation works as expected. is there side effect to this workaround ?

**Comment by mamazu** [comment]
We already have an address model the problem is that this is constructed too late. We would need to construct it in the constructor of the request or in the `fromHTTP` method of it to trigger the validation of Symfony.

**Comment by AndreasA** [comment]
You could also add a validation for the array data. It would probably be the fastest workaround.

Not that nice but it should work, see documentation:
https://symfony.com/doc/4.4/reference/constraints/Collection.html

**Comment by mamazu** [comment]
I know then the CommandProvider for this endpoint would need to be changed. Then I'd rather change the request object. But then we need to remove the validation in the Address object itself.
